### ISSUE TITLE
Provide "before" commit SHA1 for builds triggered by git push

### ISSUE THREAD

**Issue by Rondom** [appveyor/ci#1157]
We are running `git clang-format` on pull requests to detect inconsistent indentation in the changed lines.
https://github.com/dokan-dev/dokany/commit/c8ffc494482e299ed025855214db839a561ceab7

For pull-requests it is easy to determine the baseline commit to compare against (the parent of the merge-commit belonging to the branch merged into). For normal repository pushes AppVeyor does not provide an environment variable with the "previous commit" i.e. the commit before the push.

This affects the following:
* Starting build via API (for custom push hooks that call the API): https://www.appveyor.com/docs/api/projects-builds/#start-build-of-specific-branch-commit
* `before`/`old` in the webhook processing of [GitHub](https://developer.github.com/v3/activity/events/types/#pushevent ), [GitLab](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/web_hooks/web_hooks.md#push-events), [BitBucket](https://confluence.atlassian.com/bitbucket/event-payloads-740262817.html#EventPayloads-Push) and others

**Comment by Rondom** [comment]
This is also useful in other use cases: Check if a reported build message/warning has been there in the previous build of this branch, i.e. check if the current push introduced this warning.

**Comment by FeodorFitsner** [comment]
By "previous commit" you understand commit ID of the last build?

**Comment by Rondom** [comment]
> By "previous commit" you understand commit ID of the last build?

No, the commit-id "before the push" i.e. the tip of the branch before the push. See the links I posted to the GitHub, GitLab and BitBucket documentation on how to get it.
1. I push a new branch with an initial commit `A` to my remote
2. I commit three new commits, `B`, `C `and `D`
3. I push to the remote again. My remote branch would look like this: `A-B-C-D`. The `before` / `old` commit-id would be `A`

The commit-id of the last build can be useful in some situations as well I guess. I have yet to check whether it would be difficult to get "last build of that branch" via the API, but as I said this issue is about something else.

**Comment by FeodorFitsner** [comment]
So, basically it's `before` SHA from push webhook payload https://developer.github.com/v3/activity/events/types/#pushevent, right?

**Comment by Rondom** [comment]
> So, basically it's `before` SHA from push webhook payload https://developer.github.com/v3/activity/events/types/#pushevent, right?

Exactly, `before` for [GitHub](https://developer.github.com/v3/activity/events/types/#pushevent ) and [GitLab](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/doc/web_hooks/web_hooks.md#push-events), `old.target.hash` for [BitBucket](https://confluence.atlassian.com/bitbucket/event-payloads-740262817.html#EventPayloads-Push) and `refUpdates.oldObjectId` for [Visual Studio Online](https://www.visualstudio.com/en-us/docs/integrate/get-started/service-hooks/events#git.push) (double-check for the latter two) and probably for those with custom push hooks that call the API) [a new parameter in the API](https://www.appveyor.com/docs/api/projects-builds/#start-build-of-specific-branch-commit).

I use mostly GitHub and GitLab, so I do not have big need for the others, but I did the research for the others anyways for the sake of consistency.

**Comment by pkch** [comment]
And in addition, (I am 99% sure that) AppVeyor uses that value to determine what diff to check in the [file-based commit filtering](https://www.appveyor.com/docs/how-to/filtering-commits/#commit-files-github-and-bitbucket-only). It's good to expose it to API in case we need to validate that the commit filtering does what we expect it to do.

**Comment by huafu** [comment]
@pkch nope, file-based commit filtering is using only the last commit of the push badly...

> the files modified in push’s head commit

**Comment by Rondom** [comment]
> @pkch nope, file-based commit filtering is using only the last commit of the push badly...

I am not using that feature, but wouldn't that be a bug? Say, I push two commits, commit A changes files that should trigger the build, commit B (HEAD commit) changes only files that don't trigger builds. Result is that the build does not run, even though it should have been run.

**Comment by huafu** [comment]
yes that's it! and it's a known thing, explained in the doc
